File: leetcode-402.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        t = 0
        while k > 0:
            while(t < len(num)-1 and num[t] <= num[t+1]):
                t += 1
            if t < len(num)-1:
                k -= 1
                num = num[:t]+num[t+1:]
                t = max(0, t-1)
            else:
                break
            while len(num)> 0 and num[0] == '0':
                num = num[1:]
        while k > 0 and len(num)>0:
            tmp = len(num)-1
            while tmp > 0 and num[tmp] == '0':
                tmp -= 1
            k -= 1
            num = num[:tmp] + num[tmp+1:]
        if len(num) == 0:
            num = '0'
        return num
        

if __name__ == "__main__":
    so = Solution()
    print(so.removeKdigits("5337", 2))

File: leetcode/leetcode-402.cpp
/*
 *@File Name          :leetcode-402.cpp
 *@Author             :cc
 *@Version            :1.0.0
 *@Date               :2022/8/3 10:09
 *@Description        :
 *@Function List      :
 *@History            :
 */

#include "algorithm"
#include "stack"
#include "string"
using namespace std;

class Solution {
 public:
  string removeKdigits(string num, int k) {
    stack<char> sta;
    for (auto& x : num) {
      if (k <= 0 || sta.empty() || sta.top() <= x) {
        sta.push(x);
      } else {
        while (k > 0 && !sta.empty() && sta.top() > x) {
          sta.pop();
          k--;
        }
        sta.push(x);
      }
    }
    while (k > 0 && !sta.empty()) {
      sta.pop();
      k--;
    }
    string ans;
    while (!sta.empty()) {
      ans += sta.top();
      sta.pop();
    }
    while (!ans.empty() && ans.back() == '0') ans.pop_back();
    if (ans.empty()) ans = "0";
    std::reverse(ans.begin(), ans.end());
    return ans;
  }
};